packages/job-queue-plugin/src/bullmq/job-list-index.service.ts

```diff
@@ -0,0 +1,312 @@
+        this.setupEventListeners();
+        void this.migrateExistingJobs();
+    }
+
+    private setupEventListeners() {
+        if (!this.queueEvents || !this.queue) return;
+
+        // When a job is added to the queue
+        this.queueEvents.on('waiting', ({ jobId }) => {
+            void this.updateJobIndex(jobId, 'wait');
+        });
+
+        // When a job starts processing
+        this.queueEvents.on('active', ({ jobId }) => {
+            void this.updateJobIndex(jobId, 'active');
+        });
+
+        // When a job completes successfully
+        this.queueEvents.on('completed', ({ jobId }) => {
+            void this.updateJobIndex(jobId, 'completed');
+        });
+
+        // When a job fails
+        this.queueEvents.on('failed', ({ jobId }) => {
+            void this.updateJobIndex(jobId, 'failed');
+        });
+
+        // When a job is delayed
+        this.queueEvents.on('delayed', ({ jobId }) => {
+            void this.updateJobIndex(jobId, 'delayed');
+        });
+
+        // When a job is removed
+        this.queueEvents.on('removed', ({ jobId }) => {
+            void this.removeJobFromAllIndices(jobId);
+        });
+    }
+
+    /**
+     * When a job's state changes, we need to update the indexed set
+     * to reflect the new state of the job.
+     */
+    private async updateJobIndex(jobId: string, state: string) {
+        if (!this.redis || !this.queue) return;
+
+        try {
+            const job: Job | undefined = await this.queue.getJob(jobId);
+            if (!job) return;
+
+            const prefix = getPrefix(this.options);
+            const queuePrefix = `${prefix}:${this.queue.name}:`;
+            const timestamp = job.timestamp;
+
+            // Remove from all state indices first
+            await this.removeJobFromAllIndices(jobId);
+
+            // Add to the specific state index
+            const indexedKey = `${queuePrefix}queue:${job.name}:${state}` as const;
+            Logger.debug(`Adding job ${jobId} to indexed key: ${indexedKey}`, loggerCtx);
+            await this.redis.zadd(indexedKey, timestamp, jobId);
+        } catch (err: unknown) {
+            const error = err as Error;
+            Logger.error(`Failed to update job index: ${error.message}`, loggerCtx);
+        }
+    }
+
+    private async removeJobFromAllIndices(jobId: string) {
+        if (!this.redis || !this.queue) return;
+
+        try {
+            const prefix = this.options.workerOptions?.prefix ?? 'bull';
+            const queuePrefix = `${prefix}:${this.queue.name}:`;
+            const job: Job | undefined = await this.queue.getJob(jobId);
+            if (!job) return;
+
+            const states = ['wait', 'active', 'completed', 'failed', 'delayed'] as const;
+            const pipeline = this.redis.pipeline();
+
+            for (const state of states) {
+                const indexedKey = `${queuePrefix}queue:${job.name}:${state}` as const;
```

Comment:
Personal opinion:

- I'd move the `states` to an enum because theyre being reused blindly when setting up the event listeners above and migrating jobs below. That would also make the second argument of `updateJobIndex` typesafe
- The construction of `indexedKey` is also being reused by `updateJobIndex` and others, a small helper would make future changes less error prone. This is being used on the Lua side as well though, they must stay in sync as well

Reply:
Good points, I cleaned up these points and also realized we don't even need the `add()` method anymore because we automatically get that functionality by using the event listeners.